### PR TITLE
Fix modules with same name as builtins causing ICE (#3315)

### PR DESCRIPTION
**Pull request by liamnaddell**
gcc/rust/ChangeLog:

	* resolve/rust-forever-stack.h: Add a dedicated prelude node for the Language prelude
	* resolve/rust-forever-stack.hxx: Add support code for the prelude node
	* resolve/rust-late-name-resolver-2.0.cc (Late::visit): Move language prelude builtins to the prelude context
	* resolve/rust-name-resolution-context.cc: Add code for handling the prelude corner case
	* resolve/rust-rib.h: Add a special Prelude rib type

gcc/testsuite/ChangeLog:

	* rust/compile/issue-3315-1.rs: Add test for module with same name as builtin
	* rust/compile/issue-3315-2.rs: Test with utilization of i32 type
	* rust/compile/nr2/exclude: issue-3315-2.rs Does not work with NR2.0

Thank you for making Rust GCC better!

If your PR fixes an issue, you can add "Fixes #issue_number" into this
PR description and the git commit message. This way the issue will be
automatically closed when your PR is merged. If your change addresses
an issue but does not fully fix it please mark it as "Addresses #issue_number"
in the git commit message.

Here is a checklist to help you with your PR.

- \[ ] GCC development requires copyright assignment or the Developer's Certificate of Origin sign-off, see https://gcc.gnu.org/contribute.html or https://gcc.gnu.org/dco.html
- \[ ] Read contributing guidlines
- \[ ] `make check-rust` passes locally
- \[ ] Run `clang-format`
- \[ ] Added any relevant test cases to `gcc/testsuite/rust/`

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.
---

*Please write a comment explaining your change. This is the message
that will be part of the merge commit.

### CODE REVIEW

File: gcc/rust/resolve/rust-late-name-resolver-2.0.cc [hunk 1]
```diff
@@ -91,16 +91,22 @@ Late::setup_builtin_types ()
...
+  // There's a special Rib for putting prelude items, since prelude items need
+  // to satisfy certain special rules.
+  ctx.scoped (Rib::Kind::Prelude, 0, [this, &ty_ctx] (void) -> void {
+    for (const auto &builtin : builtins)
+      {
+	// we should be able to use `insert_at_root` or `insert` here, since
+	// we're at the root :) hopefully! Globbed declaration allows
+	// user-defined items in the `types` namespace
+	//  to shadow builtins. Probably unwise, but allowed by rustc!
+	auto ok = ctx.types.insert_globbed (builtin.name, builtin.node_id);
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-late-name-resolver-2.0.cc:103]: Since it's in a separate rib, `insert`, `insert_shadowable`, or `insert_globbed` would all work fine here

File: gcc/rust/resolve/rust-name-resolution-context.cc [hunk 2]
```diff
@@ -107,6 +107,11 @@ NameResolutionContext::scoped (Rib::Kind rib_kind, NodeId id,
 			       std::function<void (void)> lambda,
 			       tl::optional<Identifier> path)
 {
+  // Prelude doesn't have an access path
+  // TODO: Assert current cursor is at the root
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-name-resolution-context.cc:111]: This should already hit an assertion, so you could remove the comment or add another assertion

File: gcc/rust/resolve/rust-name-resolution-context.cc [hunk 3]
```diff
@@ -107,6 +107,11 @@ NameResolutionContext::scoped (Rib::Kind rib_kind, NodeId id,
...
+  if (rib_kind == Rib::Kind::Prelude)
+    rust_assert (!path);
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-name-resolution-context.cc:113]: This could be moved to `push_inner` or copied there, but an assertion like this should probably be in `push_inner`

**Comment by powerboat9** [comment]
> @powerboat9 I can update to move the language prelude to rust-forever-stack.hxx if you still want the change

It's alright, it looks good

**Comment by liamnaddell** [comment]
@powerboat9 Thank you for the reviews, and sorry about the TODO: comments -- I forgot about those. 

I think I fixed all of the issues you pointed out.

**Review by powerboat9 — APPROVED** [review]

**Review by philberty — COMMENTED** [review]
Looks like a regression for:

Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/borrowck/borrowck.exp ...
Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/compile.exp ...
Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/builtin/builtin_macro.exp ...
Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/mbe/mbe_macro.exp ...
Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/proc/proc_macro.exp ...
Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/nr2/compile.exp ...
FAIL: /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/mbe/macro54.rs: nr2 failure: FAIL: rust/compile/macros/mbe/macro54.rs -frust-name-resolution-2.0 (test for excess errors)

**Comment by liamnaddell** [comment]
> Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/mbe/mbe_macro.exp ...
> Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/proc/proc_macro.exp ...
> Running /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/nr2/compile.exp ...
> FAIL: /home/runner/work/gccrs/gccrs/gcc/testsuite/rust/compile/macros/mbe/macro54.rs: nr2 failure: FAIL: rust/compile/macros/mbe/macro54.rs -frust-name-resolution-2.0 (test for excess errors)

For some reason after rebasing these fail now. I'll take a look at these

**Comment by liamnaddell** [comment]
I think the strategy I chose to implement this doesn't work, since the test involves resolving something like u32::VALUE.

I'll try to re-implement

**Comment by liamnaddell** [comment]
@powerboat9 @philberty I added some code to get_segments to handle the prelude rib as a special case. I think my PR should work now

File: gcc/rust/resolve/rust-forever-stack.hxx [hunk 4]
```diff
@@ -63,6 +70,17 @@ template <Namespace N>
 void
 ForeverStack<N>::push_inner (Rib rib, Link link)
 {
+  if (rib.kind == Rib::Kind::Prelude)
+    {
+      // If you push_inner into the prelude from outside the root, you will pop
+      // back into the
+      //  root, which could screw up a traversal.
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-forever-stack.hxx:77]: Should probably adjust the formatting on this comment

File: gcc/rust/resolve/rust-forever-stack.hxx [hunk 5]
```diff
@@ -487,6 +519,7 @@ ForeverStack<N>::resolve_segments (
        *
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-forever-stack.hxx:519]: Should probably add an entry to the numbered list in the above comment, mentioning changing `searched_prelude` from `false` to `true`

File: gcc/rust/resolve/rust-forever-stack.h [hunk 6]
```diff
@@ -750,7 +754,15 @@ template <Namespace N> class ForeverStack
...
   void update_cursor (Node &new_cursor);
 
+  /* The forever stack's actual nodes */
   Node root;
+  /*
+   * A special prelude node used currently for resolving language builtins
+   * It has the root node as a parent, and acts as a "special case" for name
+   * resolution
+   */
+  Node prelude;
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-forever-stack.h:764]: I probably should have noticed and commented on this earlier, but this should really be called something like `language_prelude`, given that we'll have to deal with multiple preludes

File: gcc/rust/resolve/rust-name-resolution-context.cc [hunk 7]
```diff
@@ -107,6 +107,7 @@ NameResolutionContext::scoped (Rib::Kind rib_kind, NodeId id,
 			       std::function<void (void)> lambda,
 			       tl::optional<Identifier> path)
 {
+  // NOTE: You should be at the root node when calling this function.
```
Comment (powerboat9) [inline gcc/rust/resolve/rust-name-resolution-context.cc:110]: You should be at the root node, if pushing a `Prelude` rib kind

**Review by powerboat9 — COMMENTED** [review]
Otherwise, looks great

**Comment by powerboat9** [comment]
It looks like rustc has some [`backtracking`](https://github.com/rust-lang/rust/blob/1.49.0/compiler/rustc_resolve/src/lib.rs#L1612) capability. We may have to implement something like that later, for full parity, but that's probably outside the scope of this PR

File: gcc/rust/resolve/rust-forever-stack.h (hunk 6 above)
Comment (liamnaddell) [reply gcc/rust/resolve/rust-forever-stack.h:764]: I considered that as well, I figured that we would also dump the std::prelude and core::prelude in here as well

File: gcc/rust/resolve/rust-name-resolution-context.cc (hunk 7 above)
Comment (liamnaddell) [reply gcc/rust/resolve/rust-name-resolution-context.cc:110]: :facepalm:

Comment (liamnaddell) [reply gcc/rust/resolve/rust-name-resolution-context.cc:110]: I'll update to clarify that we only need to be at the root for the *prelude* rib pushes

File: gcc/rust/resolve/rust-forever-stack.h (hunk 6 above)
Comment (powerboat9) [reply gcc/rust/resolve/rust-forever-stack.h:764]: I figured we'd keep some sort of reference to the standard library prelude -- since that should already be a module with its own `Node`, we shouldn't need to try to copy entries into another `Node`

File: gcc/rust/resolve/rust-forever-stack.hxx (hunk 4 above)
Comment (liamnaddell) [reply gcc/rust/resolve/rust-forever-stack.hxx]: fixed

File: gcc/rust/resolve/rust-forever-stack.hxx (hunk 5 above)
Comment (liamnaddell) [reply gcc/rust/resolve/rust-forever-stack.hxx:517]: I added a comment for this, let me know if it needs modification.

**Review by powerboat9 — APPROVED** [review]

File: gcc/rust/resolve/rust-forever-stack.h (hunk 6 above)
Comment (liamnaddell) [reply gcc/rust/resolve/rust-forever-stack.h:764]: I'll add a comment to clarify this

**Comment by dkm** [comment]
Small note, your changelog is not correct. It should not only give the file name, but also the "item" it is modifying. Usually, a function, but can be a type, or something else.

```
   * file.c (function_name): Adjusted call to bla
```

**Comment by liamnaddell** [comment]
> Small note, your changelog is not correct. It should not only give the file name, but also the "item" it is modifying. Usually, a function, but can be a type, or something else.
> 
> ```
>    * file.c (function_name): Adjusted call to bla
> ```



> Small note, your changelog is not correct. It should not only give the file name, but also the "item" it is modifying. Usually, a function, but can be a type, or something else.

@dkm Is the log better now?

**Review by P-E-P — APPROVED** [review]
I'm unsure this is the correct approach, but I can't find any satisfying way to solve this problem for now so let's keep it as is. Code-wise I've got nothing to say, LGTM.
